File: apps/agent/src/state.py
from typing import Optional, Any
from copilotkit import CopilotKitState


class AgentState(CopilotKitState):
    """
    Liquid agent state.

    Extends CopilotKitState (which includes messages + copilotkit fields)
    with the four domain-specific fields driven entirely by frontend controls.
    """
    inputText: str
    controls: Optional[Any]     # ControlSchema serialised as dict, or None
    activeValues: dict           # { [controlId]: number | boolean }
    outputText: str
    sessionId: str

File: apps/agent/src/state.ts
import { Annotation } from "@langchain/langgraph";
import { CopilotKitStateAnnotation } from "@copilotkit/sdk-js/langgraph";
import type { ControlSchema, ActiveValues } from "./types";

export const AgentStateAnnotation = Annotation.Root({
  ...CopilotKitStateAnnotation.spec,
  // Use Annotation<T> without options — matches the LangGraph starter pattern.
  // "last write wins" reducer; initial values come from useCoAgent's initialState.
  inputText:    Annotation<string>,
  controls:     Annotation<ControlSchema | null>,
  activeValues: Annotation<ActiveValues>,
  outputText:   Annotation<string>,
  sessionId:    Annotation<string>,
});

export type AgentState = typeof AgentStateAnnotation.State;
